src/core/gui/sidebar/previews/base/SidebarPreviewBaseEntry.cpp

```diff
@@ -132,6 +132,41 @@ void SidebarPreviewBaseEntry::paint(cairo_t* cr) {
                            round_cast<int>(width), round_cast<int>(height));
     }
 
+    std::optional<Recolor> recolor;
+    {
+        auto r = sidebar->getControl()->getSettings()->getRecolor();
+        recolor = std::get<1>(r) ? std::make_optional(std::get<2>(r)) : std::nullopt;
+    }
+    // recoloring is done by
+    // 1. invert the whole background
+    // 2. scale (multiply) the color spectrum to fit between low/dark and high/light
+    // 3. move (add) the color spectrum to lay exactly in range low/dark and high/light
+    if (recolor) {
+        // encapsulate in save/restore to limit the scope of the clip operation
+        cairo_save(cr);
+
+        cairo_rectangle(cr, Shadow::getShadowTopLeftSize() + 2, Shadow::getShadowTopLeftSize() + 2, width, height);
+        // constrain the area which is painted on
+        cairo_clip(cr);
+
+        // Apply inversion
+        cairo_set_operator(cr, CAIRO_OPERATOR_DIFFERENCE);
+        cairo_set_source_rgb(cr, 1., 1., 1.);
+        cairo_paint(cr);
+
+        // Scale the spectrum
+        cairo_set_operator(cr, CAIRO_OPERATOR_MULTIPLY);
+        Util::cairo_set_source_rgbi(cr, recolor->getDifference());
+        cairo_paint(cr);
+
+        // Move the spectrum
+        cairo_set_operator(cr, CAIRO_OPERATOR_ADD);
+        Util::cairo_set_source_rgbi(cr, recolor->getOffset());
+        cairo_paint(cr);
```

Comment:
Yea, I thought about it. Apart from saving calculations, this was the idea of moving the `diff` and `offset` to `Recolor.cpp`. I was just a bit unsure of putting "raw" `cairo` calls in `Recolor.cpp`. Especially since it needs to be guarded by `cairo_save`/`cairo_restore` and `cairo_clip` is involved in the other case.

But in this case I think it's possible and it definitely increases the code quality.